gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeCatalog.java

```diff
@@ -0,0 +1,398 @@
+      // that works on GCP services (e.g., GCE, GKE, Dataproc).
+      newClient =
+          new BigLakeClient(
+              properties.getOrDefault(
+                  PROPERTIES_KEY_BIGLAKE_ENDPOINT, DEFAULT_BIGLAKE_SERVICE_ENDPOINT),
+              propProjectId,
+              propLocation);
+    } catch (IOException e) {
+      throw new ServiceFailureException(e, "Creating BigLake client failed");
+    }
+    initialize(inputName, properties, propProjectId, propLocation, newClient);
+  }
+
+  @VisibleForTesting
+  void initialize(
+      String inputName,
+      Map<String, String> properties,
+      String propProjectId,
+      String propLocation,
+      BigLakeClient bigLakeClient) {
+    this.catalogPulginName = inputName;
+    this.catalogProperties = ImmutableMap.copyOf(properties);
+    this.projectId = propProjectId;
+    this.location = propLocation;
+    Preconditions.checkNotNull(bigLakeClient, "BigLake client must not be null");
+    this.client = bigLakeClient;
+
+    // Users can specify the BigLake catalog ID, otherwise catalog plugin will be used.
+    this.catalogId = properties.getOrDefault(PROPERTIES_KEY_BLMS_CATALOG, inputName);
+    this.catalogName = CatalogName.of(projectId, location, catalogId);
+    LOG.info("Use BigLake catalog: {}", catalogName.toString());
+
+    String fileIOImpl =
+        properties.getOrDefault(CatalogProperties.FILE_IO_IMPL, ResolvingFileIO.class.getName());
+    this.fileIO = CatalogUtil.loadFileIO(fileIOImpl, properties, conf);
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier identifier) {
+    // The identifier of metadata tables is like "ns.table.files".
+    // We return a non-existing table in this case (empty table ID is disallowed in BigLake
+    // Metastore), loadTable will try loadMetadataTable.
+    if (identifier.namespace().levels().length > 1
+        && MetadataTableType.from(identifier.name()) != null) {
+      return new BigLakeTableOperations(
+          client, fileIO, getTableName(identifier.namespace().level(0), /* tableId= */ ""));
+    }
+
+    return new BigLakeTableOperations(
+        client,
+        fileIO,
+        getTableName(getDatabaseId(identifier.namespace()), /* tableId= */ identifier.name()));
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier identifier) {
+    String locationUri = getDatabase(identifier.namespace()).getHiveOptions().getLocationUri();
+    return String.format(
+        "%s/%s",
+        Strings.isNullOrEmpty(locationUri)
+            ? getDatabaseLocation(getDatabaseId(identifier.namespace()))
+            : locationUri,
+        identifier.name());
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    // When deleting a BLMS catalog via `DROP NAMESPACE <catalog>`, this method is called for
+    // verifying catalog emptiness. `namespace` is empty in this case, we list databases in
+    // this catalog instead.
```

Comment:
Can you point me to where this is happening? I don't think it is right for this to list databases. If tables can't be tracked directly in the empty namespace, then this should always return `ImmutableList.of()`.

Reply:
It happens in an older Spark version: https://github.com/apache/spark/blob/2f3e4e36017d16d67086fd4ecaf39636a2fb4b7c/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala#L253, the listTables check is now removed.

I list all tables in the catalog (all databases) instead.